File: app/src/main/java/pl/szczodrzynski/edziennik/messages/MessagesComposeSuggestionAdapter.java
package pl.szczodrzynski.edziennik.messages;

import android.content.Context;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import pl.szczodrzynski.edziennik.R;
import pl.szczodrzynski.edziennik.datamodels.Teacher;

public class MessagesComposeSuggestionAdapter extends ArrayAdapter<Teacher> {

    private Context context;
    private List<Teacher> teacherList;
    private ArrayList<Teacher> originalList = null;
    private ArrayFilter mFilter;
    private final Object mLock = new Object();

    MessagesComposeSuggestionAdapter(@NonNull Context context, List<Teacher> teacherList) {
        super(context, 0, teacherList);
        this.context = context;
        this.teacherList = teacherList;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        View listItem = convertView;
        if(listItem == null)
            listItem = LayoutInflater.from(context).inflate(R.layout.messages_compose_suggestion_item, parent, false);

        Teacher teacher = teacherList.get(position);

        TextView name = listItem.findViewById(R.id.name);
        TextView type = listItem.findViewById(R.id.type);
        ImageView image = listItem.findViewById(R.id.image);

        teacher.image = MessagesUtils.getProfileImage(48, 24, 16, 12, 1, teacher.getFullName());

        if (teacher.id <= 0) {
            name.setText(Teacher.typeString(context, (int) (teacher.id * -1)));
            type.setText(R.string.teachers_browse_category);
            image.setImageBitmap(null);
        }
        else {
            if (teacher.displayName == null)
                name.setText(teacher.getFullName());
            else
                name.setText(Html.fromHtml(teacher.displayName));
            type.setText(teacher.getType(context));
            image.setImageBitmap(teacher.image);
        }

        return listItem;
    }

    @Override
    public int getCount() {
        return teacherList.size();
    }

    @Nullable
    @Override
    public Teacher getItem(int position) {
        return teacherList.get(position);
    }

    @Override
    public int getPosition(@Nullable Teacher item) {
        return teacherList.indexOf(item);
    }

    @NonNull
    @Override
    public Filter getFilter() {
        if (mFilter == null) {
            mFilter = new ArrayFilter();
        }
        return mFilter;
    }

    private class TeacherWeighted extends Teacher {
        public int weight;

        public TeacherWeighted(Teacher teacher, int weight) {
            super(teacher.profileId, teacher.id, teacher.name, teacher.surname, teacher.loginId);
            this.weight = weight;
            this.image = teacher.image;
            this.type = teacher.type;
            this.typeDescription = teacher.typeDescription;
        }

        @NonNull
        @Override
        public String toString() {
            return getFullName();
        }
    }
    private Comparator<? super Teacher> comparator = (o1, o2) -> ((TeacherWeighted) o1).weight - ((TeacherWeighted) o2).weight;

    private class ArrayFilter extends Filter {
        @Override
        protected FilterResults performFiltering(CharSequence prefix) {
            final FilterResults results = new FilterResults();

            if (originalList == null) {
                synchronized (mLock) {
                    originalList = new ArrayList<>(teacherList);
                }
            }

            if (prefix == null || prefix.length() == 0) {
                ArrayList<Teacher> list;
                synchronized (mLock) {
                    list = new ArrayList<>(originalList);
                }
                results.values = list;
                results.count = list.size();
            } else {
                String prefixString = prefix.toString().toLowerCase();

                ArrayList<Teacher> values;
                synchronized (mLock) {
                    values = new ArrayList<>(originalList);
                }

                int count = values.size();
                ArrayList<Teacher> newValues = new ArrayList<>();

                for (int i = 0; i < count; i++) {
                    Teacher teacher = values.get(i);
                    String teacherFullName = teacher.getFullName().toLowerCase();
                    teacher.displayName = teacherFullName.replace(prefixString, "<b>"+prefixString+"</b>");

                    // First match against the whole, non-splitted value
                    boolean found = false;
                    if (teacherFullName.startsWith(prefixString)) {
                        newValues.add(new TeacherWeighted(teacher, 1));
                        found = true;
                    } else {
                        // check if prefix matches any of the words
                        String[] words = teacherFullName.split(" ");
                        for (String word : words) {
                            if (word.startsWith(prefixString)) {
                                newValues.add(new TeacherWeighted(teacher, 2));
                                found = true;
                                break;
                            }
                        }
                    }
                    // finally check if the prefix matches any part of the name
                    if (!found && teacherFullName.contains(prefixString)) {
                        newValues.add(new TeacherWeighted(teacher, 3));
                    }

                }

                Collections.sort(newValues, comparator);

                results.values = newValues;
                results.count = newValues.size();
            }

            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            //noinspection unchecked
            teacherList = (List<Teacher>) results.values;
            if (results.count > 0) {
                notifyDataSetChanged();
            } else {
                notifyDataSetInvalidated();
            }
        }
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/modules/messages/MessagesComposeSuggestionAdapter.kt
package pl.szczodrzynski.edziennik.ui.modules.messages

import android.content.Context
import android.graphics.Typeface.BOLD
import android.text.style.BackgroundColorSpan
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.ImageView
import android.widget.TextView
import pl.szczodrzynski.edziennik.R
import pl.szczodrzynski.edziennik.asSpannable
import pl.szczodrzynski.edziennik.data.db.entity.Teacher
import pl.szczodrzynski.edziennik.resolveAttr
import pl.szczodrzynski.edziennik.ui.modules.messages.MessagesUtils.getProfileImage
import java.util.*

class MessagesComposeSuggestionAdapter(
        context: Context,
        val originalList: List<Teacher>
) : ArrayAdapter<Teacher>(context, 0, originalList)
{

    private var teacherList = originalList.toList()
    private val filter by lazy { ArrayFilter() }
    private val comparator by lazy { Comparator { o1: Teacher, o2: Teacher -> o1.recipientWeight - o2.recipientWeight } }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val listItem = convertView ?: LayoutInflater.from(context).inflate(R.layout.messages_compose_suggestion_item, parent, false)

        val teacher = teacherList[position]
        val name = listItem.findViewById<TextView>(R.id.name)
        val type = listItem.findViewById<TextView>(R.id.type)
        val image = listItem.findViewById<ImageView>(R.id.image)

        if (teacher.image == null)
            teacher.image = getProfileImage(48, 24, 16, 12, 1, teacher.fullName)

        if (teacher.id in -24L..0L) {
            name.text = Teacher.typeName(context, (teacher.id * -1).toInt())
            type.setText(R.string.teachers_browse_category)
            image.setImageBitmap(null)
        } else {
            name.text = teacher.recipientDisplayName ?: teacher.fullName
            type.text = teacher.getTypeText(context)
            image.setImageBitmap(teacher.image)
        }
        return listItem
    }

    override fun getCount(): Int {
        return teacherList.size
    }

    override fun getItem(position: Int): Teacher? {
        return teacherList[position]
    }

    override fun getPosition(item: Teacher?): Int {
        return teacherList.indexOf(item)
    }

    override fun getFilter(): Filter {
        return filter
    }

    inner class ArrayFilter : Filter() {
        override fun performFiltering(prefix: CharSequence?): FilterResults {
            val results = FilterResults()

            if (prefix == null) {
                originalList.filter { it.id in -24L..0L }.let {
                    results.values = it
                    results.count = it.size
                }
            } else if (prefix.isEmpty()) {
                results.values = originalList
                results.count = originalList.size
            } else {
                val prefixString = prefix.toString()

                val list = mutableListOf<Teacher>()

                originalList.forEach { teacher ->
                    val teacherFullName = teacher.fullName
                    teacher.recipientWeight = 0

                    // First match against the whole, non-split value
                    var found = false
                    if (teacherFullName.startsWith(prefixString, ignoreCase = true)) {
                        teacher.recipientWeight = 1
                        found = true
                    } else {
                        // check if prefix matches any of the words
                        val words = teacherFullName.split(" ").toTypedArray()
                        for (word in words) {
                            if (word.startsWith(prefixString, ignoreCase = true)) {
                                teacher.recipientWeight = 2
                                found = true
                                break
                            }
                        }
                    }
                    // finally check if the prefix matches any part of the name
                    if (!found && teacherFullName.contains(prefixString, ignoreCase = true)) {
                        teacher.recipientWeight = 3
                    }

                    if (teacher.recipientWeight != 0) {
                        teacher.recipientDisplayName = teacherFullName.asSpannable(
                                StyleSpan(BOLD),
                                BackgroundColorSpan(R.attr.colorControlHighlight.resolveAttr(context)),
                                substring = prefixString,
                                ignoreCase = true
                        )
                        list += teacher
                    }
                }

                Collections.sort(list, comparator)
                results.values = list
                results.count = list.size
            }
            return results
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults) {
            results.values?.let { teacherList = it as List<Teacher> }
            if (results.count > 0) {
                notifyDataSetChanged()
            } else {
                notifyDataSetInvalidated()
            }
        }
    }

}
